sd/endpointer.go

```diff
@@ -0,0 +1,75 @@
+package sd
+
+import (
+	"io"
+
+	"github.com/go-kit/kit/endpoint"
+	"github.com/go-kit/kit/log"
+	iEndpoint "github.com/go-kit/kit/sd/internal/endpoint"
+)
+
+// Endpointer listens to a service discovery system and yields a set of
+// identical endpoints on demand. An error indicates a problem with connectivity
+// to the service discovery system, or within the system itself; an Endpointer
+// may yield no endpoints without error.
+type Endpointer interface {
+	Endpoints() ([]endpoint.Endpoint, error)
+}
+
+// Factory is a function that converts an instance string (e.g. host:port) to a
+// specific endpoint. Instances that provide multiple endpoints require multiple
+// factories. A factory also returns an io.Closer that's invoked when the
+// instance goes away and needs to be cleaned up. Factories may return nil
+// closers.
+//
+// Users are expected to provide their own factory functions that assume
+// specific transports, or can deduce transports by parsing the instance string.
+type Factory func(instance string) (endpoint.Endpoint, io.Closer, error)
+
+// NewEndpointer creates an Endpointer.
+func NewEndpointer(src Instancer, f Factory, logger log.Logger) Endpointer {
+	return newSimpleEndpointer(src, f, logger)
+}
+
+// FixedEndpointer yields a fixed set of endpoints.
+type FixedEndpointer []endpoint.Endpoint
+
+// Endpoints implements Endpointer.
+func (s FixedEndpointer) Endpoints() ([]endpoint.Endpoint, error) { return s, nil }
+
+func newSimpleEndpointer(src Instancer, f Factory, logger log.Logger) Endpointer {
+	// Cache declares its own Factory to avoid circular dependecy on sd package
+	factory := func(instance string) (endpoint.Endpoint, io.Closer, error) {
+		return f(instance)
+	}
+	se := &simpleEndpointer{
+		Cache:     *iEndpoint.NewCache(factory, logger),
```

Comment:
What prevents e.g. endpointCache from being defined in this package as a non-exported type?